File: sdk/python/ahoy_tokenisation/resources/webhooks.py
"""
Webhooks resource for managing webhook endpoints and deliveries.
"""

from typing import Any, Dict, List, Optional

from ahoy_tokenisation.models import WebhookEndpoint, WebhookDelivery
from ahoy_tokenisation.utils.http import HttpClient, AsyncHttpClient
from ahoy_tokenisation.utils.pagination import PaginatedList, AsyncPaginatedList


class WebhooksResource:
    """Manage webhook endpoints and deliveries."""

    def __init__(self, http: HttpClient):
        self._http = http

    def create(
        self,
        url: str,
        events: List[str],
        *,
        description: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> WebhookEndpoint:
        """
        Create a webhook endpoint.

        Args:
            url: The webhook URL
            events: List of event types to subscribe to
            description: Endpoint description
            metadata: Custom metadata

        Returns:
            Created WebhookEndpoint (includes secret on creation only)

        Example:
            endpoint = client.webhooks.create(
                url="https://your-app.com/webhooks/ahoy",
                events=["investor.created", "transfer.completed"],
            )
            # Store endpoint.secret securely!
        """
        data = {
            "url": url,
            "events": events,
        }
        if description is not None:
            data["description"] = description
        if metadata is not None:
            data["metadata"] = metadata

        response = self._http.post("/v1/webhooks", json=data)
        return WebhookEndpoint(**response)

    def retrieve(self, endpoint_id: str) -> WebhookEndpoint:
        """
        Retrieve a webhook endpoint.

        Args:
            endpoint_id: The endpoint's ID

        Returns:
            WebhookEndpoint object (secret not included)
        """
        response = self._http.get(f"/v1/webhooks/{endpoint_id}")
        return WebhookEndpoint(**response)

    def update(
        self,
        endpoint_id: str,
        *,
        url: Optional[str] = None,
        events: Optional[List[str]] = None,
        description: Optional[str] = None,
        status: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> WebhookEndpoint:
        """
        Update a webhook endpoint.

        Args:
            endpoint_id: The endpoint's ID
            url: New URL
            events: Updated event list
            description: New description
            status: New status (active, paused)
            metadata: Updated metadata

        Returns:
            Updated WebhookEndpoint
        """
        data = {}
        if url is not None:
            data["url"] = url
        if events is not None:
            data["events"] = events
        if description is not None:
            data["description"] = description
        if status is not None:
            data["status"] = status
        if metadata is not None:
            data["metadata"] = metadata

        response = self._http.patch(f"/v1/webhooks/{endpoint_id}", json=data)
        return WebhookEndpoint(**response)

    def delete(self, endpoint_id: str) -> None:
        """
        Delete a webhook endpoint.

        Args:
            endpoint_id: The endpoint's ID
        """
        self._http.delete(f"/v1/webhooks/{endpoint_id}")

    def list(
        self,
        *,
        limit: int = 20,
        offset: int = 0,
        status: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        List webhook endpoints.

        Args:
            limit: Maximum number of results
            offset: Number of results to skip
            status: Filter by status

        Returns:
            Paginated response with endpoints
        """
        params = {"limit": limit, "offset": offset}
        if status is not None:
            params["status"] = status

        response = self._http.get("/v1/webhooks", params=params)
        response["data"] = [WebhookEndpoint(**e) for e in response.get("data", [])]
        return response

    def rotate_secret(self, endpoint_id: str) -> WebhookEndpoint:
        """
        Rotate the webhook secret.

        Args:
            endpoint_id: The endpoint's ID

        Returns:
            WebhookEndpoint with new secret
        """
        response = self._http.post(f"/v1/webhooks/{endpoint_id}/rotate-secret")
        return WebhookEndpoint(**response)

    # Delivery Methods
    def list_deliveries(
        self,
        endpoint_id: str,
        *,
        limit: int = 20,
        offset: int = 0,
        status: Optional[str] = None,
        event_type: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        List webhook deliveries for an endpoint.

        Args:
            endpoint_id: The endpoint's ID
            limit: Maximum number of results
            offset: Number of results to skip
            status: Filter by delivery status
            event_type: Filter by event type

        Returns:
            Paginated response with deliveries
        """
        params = {"limit": limit, "offset": offset}
        if status is not None:
            params["status"] = status
        if event_type is not None:
            params["eventType"] = event_type

        response = self._http.get(f"/v1/webhooks/{endpoint_id}/deliveries", params=params)
        response["data"] = [WebhookDelivery(**d) for d in response.get("data", [])]
        return response

    def retry_delivery(self, endpoint_id: str, delivery_id: str) -> WebhookDelivery:
        """
        Retry a failed delivery.

        Args:
            endpoint_id: The endpoint's ID
            delivery_id: The delivery's ID

        Returns:
            Updated WebhookDelivery
        """
        response = self._http.post(f"/v1/webhooks/{endpoint_id}/deliveries/{delivery_id}/retry")
        return WebhookDelivery(**response)

    # Event Types
    def list_event_types(self) -> List[Dict[str, Any]]:
        """
        List available webhook event types.

        Returns:
            List of event type definitions
        """
        response = self._http.get("/v1/webhooks/event-types")
        return response.get("data", [])


class AsyncWebhooksResource:
    """Async version of WebhooksResource."""

    def __init__(self, http: AsyncHttpClient):
        self._http = http

    async def create(
        self,
        url: str,
        events: List[str],
        *,
        description: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> WebhookEndpoint:
        """Create a webhook endpoint."""
        data = {"url": url, "events": events}
        if description is not None:
            data["description"] = description
        if metadata is not None:
            data["metadata"] = metadata

        response = await self._http.post("/v1/webhooks", json=data)
        return WebhookEndpoint(**response)

    async def retrieve(self, endpoint_id: str) -> WebhookEndpoint:
        """Retrieve a webhook endpoint."""
        response = await self._http.get(f"/v1/webhooks/{endpoint_id}")
        return WebhookEndpoint(**response)

    async def update(
        self,
        endpoint_id: str,
        *,
        url: Optional[str] = None,
        events: Optional[List[str]] = None,
        description: Optional[str] = None,
        status: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> WebhookEndpoint:
        """Update a webhook endpoint."""
        data = {}
        if url is not None:
            data["url"] = url
        if events is not None:
            data["events"] = events
        if description is not None:
            data["description"] = description
        if status is not None:
            data["status"] = status
        if metadata is not None:
            data["metadata"] = metadata

        response = await self._http.patch(f"/v1/webhooks/{endpoint_id}", json=data)
        return WebhookEndpoint(**response)

    async def delete(self, endpoint_id: str) -> None:
        """Delete a webhook endpoint."""
        await self._http.delete(f"/v1/webhooks/{endpoint_id}")

    async def list(
        self,
        *,
        limit: int = 20,
        offset: int = 0,
        status: Optional[str] = None,
    ) -> Dict[str, Any]:
        """List webhook endpoints."""
        params = {"limit": limit, "offset": offset}
        if status is not None:
            params["status"] = status

        response = await self._http.get("/v1/webhooks", params=params)
        response["data"] = [WebhookEndpoint(**e) for e in response.get("data", [])]
        return response

    async def rotate_secret(self, endpoint_id: str) -> WebhookEndpoint:
        """Rotate the webhook secret."""
        response = await self._http.post(f"/v1/webhooks/{endpoint_id}/rotate-secret")
        return WebhookEndpoint(**response)

    async def list_deliveries(
        self,
        endpoint_id: str,
        *,
        limit: int = 20,
        offset: int = 0,
        status: Optional[str] = None,
        event_type: Optional[str] = None,
    ) -> Dict[str, Any]:
        """List webhook deliveries for an endpoint."""
        params = {"limit": limit, "offset": offset}
        if status is not None:
            params["status"] = status
        if event_type is not None:
            params["eventType"] = event_type

        response = await self._http.get(f"/v1/webhooks/{endpoint_id}/deliveries", params=params)
        response["data"] = [WebhookDelivery(**d) for d in response.get("data", [])]
        return response

    async def retry_delivery(self, endpoint_id: str, delivery_id: str) -> WebhookDelivery:
        """Retry a failed delivery."""
        response = await self._http.post(f"/v1/webhooks/{endpoint_id}/deliveries/{delivery_id}/retry")
        return WebhookDelivery(**response)

    async def list_event_types(self) -> List[Dict[str, Any]]:
        """List available webhook event types."""
        response = await self._http.get("/v1/webhooks/event-types")
        return response.get("data", [])

File: packages/core/src/modules/webhooks.ts
import type { HttpClient } from '../utils/http.js';
import type { WebhookEndpoint, WebhookDelivery, PaginatedResponse } from '../types.js';
import { validate, SendTestEventInputSchema } from './validation.js';
import { z } from 'zod';

// ============================================================================
// Validation Schemas
// ============================================================================

const EventPatternSchema = z.string().regex(/^[a-z_]+(\.[a-z_*]+)*$/, 'Invalid event pattern');

const CreateEndpointSchema = z.object({
  url: z.string().url(),
  description: z.string().optional(),
  events: z.array(EventPatternSchema).min(1),
  metadata: z.record(z.unknown()).optional(),
});

const UpdateEndpointSchema = z.object({
  url: z.string().url().optional(),
  description: z.string().optional(),
  events: z.array(EventPatternSchema).optional(),
  status: z.enum(['active', 'disabled']).optional(),
  metadata: z.record(z.unknown()).optional(),
});

const ListEndpointsParamsSchema = z.object({
  status: z.enum(['active', 'disabled']).optional(),
  limit: z.number().int().min(1).max(100).optional(),
  offset: z.number().int().min(0).optional(),
});

const ListDeliveriesParamsSchema = z.object({
  endpointId: z.string().uuid().optional(),
  status: z.enum(['pending', 'sent', 'failed', 'dlq']).optional(),
  eventType: z.string().optional(),
  limit: z.number().int().min(1).max(100).optional(),
  offset: z.number().int().min(0).optional(),
});

const UUIDSchema = z.string().uuid();

// ============================================================================
// Types
// ============================================================================

export interface CreateWebhookEndpointInput {
  /** The URL to receive webhook events (must be HTTPS in production) */
  url: string;
  /** Human-readable description */
  description?: string;
  /** Event patterns to subscribe to (e.g., ["transfer.*", "token.deployed"]) */
  events: string[];
  /** Optional metadata */
  metadata?: Record<string, unknown>;
}

export interface UpdateWebhookEndpointInput {
  /** New URL */
  url?: string;
  /** New description */
  description?: string;
  /** New event patterns */
  events?: string[];
  /** Enable or disable the endpoint */
  status?: 'active' | 'disabled';
  /** Updated metadata */
  metadata?: Record<string, unknown>;
}

export interface ListEndpointsParams {
  /** Filter by status */
  status?: 'active' | 'disabled';
  /** Maximum results */
  limit?: number;
  /** Pagination offset */
  offset?: number;
}

export interface ListDeliveriesParams {
  /** Filter by endpoint ID */
  endpointId?: string;
  /** Filter by delivery status */
  status?: 'pending' | 'sent' | 'failed' | 'dlq';
  /** Filter by event type */
  eventType?: string;
  /** Maximum results */
  limit?: number;
  /** Pagination offset */
  offset?: number;
}

export interface WebhookEndpointWithSecret extends WebhookEndpoint {
  /** Webhook signing secret (only returned on creation) */
  secret: string;
  /** Warning about storing the secret */
  warning?: string;
}

export interface TestWebhookResult {
  eventId: string;
  deliveryCount: number;
}

// ============================================================================
// Webhooks Module
// ============================================================================

/**
 * Webhooks module for managing webhook endpoints and deliveries.
 *
 * Webhooks enable real-time notifications when events occur in your organization.
 *
 * Features:
 * - HMAC-SHA256 signed payloads for security
 * - Automatic retries with exponential backoff
 * - Event pattern matching (e.g., "transfer.*" matches all transfer events)
 * - Delivery tracking and retry capabilities
 *
 * @example
 * ```typescript
 * // Create a webhook endpoint
 * const endpoint = await client.webhooks.create({
 *   url: 'https://api.example.com/webhooks',
 *   events: ['transfer.*', 'token.deployed'],
 *   description: 'Production webhook'
 * });
 *
 * // Save the secret for verifying signatures
 * console.log('Webhook secret:', endpoint.secret);
 * ```
 */
export class WebhooksModule {
  constructor(private http: HttpClient) {}

  // ============================================================================
  // Endpoint Management
  // ============================================================================

  /**
   * Creates a new webhook endpoint.
   *
   * **Important**: Save the returned `secret` securely. It will not be shown again
   * and is required to verify webhook signatures.
   *
   * @example
   * ```typescript
   * const endpoint = await client.webhooks.create({
   *   url: 'https://api.example.com/webhooks',
   *   events: ['transfer.*', 'investor.activated'],
   *   description: 'Main webhook endpoint'
   * });
   *
   * // Store this secret securely!
   * saveToVault(endpoint.id, endpoint.secret);
   * ```
   */
  async create(input: CreateWebhookEndpointInput): Promise<WebhookEndpointWithSecret> {
    const validated = validate(CreateEndpointSchema, input);
    const response = await this.http.post<WebhookEndpointWithSecret>('/api/v1/webhooks/endpoints', validated);
    return response.data;
  }

  /**
   * Retrieves a webhook endpoint by ID.
   * Note: The secret is not returned for security.
   */
  async get(id: string): Promise<WebhookEndpoint> {
    const validatedId = validate(UUIDSchema, id);
    const response = await this.http.get<WebhookEndpoint>(`/api/v1/webhooks/endpoints/${validatedId}`);
    return response.data;
  }

  /**
   * Lists all webhook endpoints.
   *
   * @example
   * ```typescript
   * // Get all active endpoints
   * const endpoints = await client.webhooks.list({ status: 'active' });
   * ```
   */
  async list(params?: ListEndpointsParams): Promise<PaginatedResponse<WebhookEndpoint>> {
    const validated = params ? validate(ListEndpointsParamsSchema, params) : undefined;
    return this.http.list<WebhookEndpoint>(
      '/api/v1/webhooks/endpoints',
      validated as Record<string, string | number | boolean | undefined>
    );
  }

  /**
   * Updates a webhook endpoint.
   *
   * @example
   * ```typescript
   * // Disable an endpoint
   * await client.webhooks.update(endpointId, { status: 'disabled' });
   *
   * // Change subscribed events
   * await client.webhooks.update(endpointId, {
   *   events: ['transfer.confirmed', 'transfer.settled']
   * });
   * ```
   */
  async update(id: string, input: UpdateWebhookEndpointInput): Promise<WebhookEndpoint> {
    const validatedId = validate(UUIDSchema, id);
    const validated = validate(UpdateEndpointSchema, input);
    const response = await this.http.patch<WebhookEndpoint>(
      `/api/v1/webhooks/endpoints/${validatedId}`,
      validated
    );
    return response.data;
  }

  /**
   * Deletes a webhook endpoint.
   * This also cancels any pending deliveries for this endpoint.
   */
  async delete(id: string): Promise<{ success: boolean; id: string }> {
    const validatedId = validate(UUIDSchema, id);
    const response = await this.http.delete<{ success: boolean; id: string }>(
      `/api/v1/webhooks/endpoints/${validatedId}`
    );
    return response.data;
  }

  /**
   * Rotates the signing secret for a webhook endpoint.
   *
   * **Important**: After rotation, update your webhook receiver to use the new secret.
   * The old secret will immediately stop working.
   *
   * @example
   * ```typescript
   * const result = await client.webhooks.rotateSecret(endpointId);
   * console.log('New secret:', result.secret);
   * // Update your webhook receiver configuration
   * ```
   */
  async rotateSecret(id: string): Promise<{ secret: string; warning: string }> {
    const validatedId = validate(UUIDSchema, id);
    const response = await this.http.post<{ secret: string; warning: string }>(
      `/api/v1/webhooks/endpoints/${validatedId}/rotate-secret`
    );
    return response.data;
  }

  // ============================================================================
  // Deliveries
  // ============================================================================

  /**
   * Lists webhook deliveries with optional filters.
   *
   * @example
   * ```typescript
   * // Get failed deliveries
   * const failed = await client.webhooks.listDeliveries({ status: 'failed' });
   *
   * // Get deliveries for a specific endpoint
   * const deliveries = await client.webhooks.listDeliveries({
   *   endpointId: 'endpoint-id',
   *   limit: 50
   * });
   * ```
   */
  async listDeliveries(params?: ListDeliveriesParams): Promise<PaginatedResponse<WebhookDelivery>> {
    const validated = params ? validate(ListDeliveriesParamsSchema, params) : undefined;
    return this.http.list<WebhookDelivery>(
      '/api/v1/webhooks/deliveries',
      validated as Record<string, string | number | boolean | undefined>
    );
  }

  /**
   * Retries a failed webhook delivery.
   *
   * @example
   * ```typescript
   * const result = await client.webhooks.retryDelivery(deliveryId);
   * if (result.success) {
   *   console.log('Delivery successful');
   * } else {
   *   console.log('Delivery failed:', result.error);
   * }
   * ```
   */
  async retryDelivery(deliveryId: string): Promise<{ success: boolean; error?: string }> {
    const validatedId = validate(UUIDSchema, deliveryId);
    const response = await this.http.post<{ success: boolean; error?: string }>(
      `/api/v1/webhooks/deliveries/${validatedId}/retry`
    );
    return response.data;
  }

  // ============================================================================
  // Testing
  // ============================================================================

  /**
   * Sends a test event to all matching webhook endpoints.
   * Useful for verifying webhook configuration.
   *
   * @example
   * ```typescript
   * // Test with a custom event
   * const result = await client.webhooks.sendTestEvent('test.event', {
   *   message: 'This is a test'
   * });
   * console.log(`Sent to ${result.deliveryCount} endpoints`);
   * ```
   */
  async sendTestEvent(
    eventType: string,
    data?: Record<string, unknown>
  ): Promise<TestWebhookResult> {
    const validated = validate(SendTestEventInputSchema, { eventType, data });
    const response = await this.http.post<TestWebhookResult>('/api/v1/webhooks/test', {
      eventType: validated.eventType,
      data: validated.data || { test: true, timestamp: new Date().toISOString() }
    });
    return response.data;
  }

  // ============================================================================
  // Signature Verification (Static Helper)
  // ============================================================================

  /**
   * Verifies a webhook signature.
   * Use this in your webhook receiver to validate that the request came from us.
   *
   * @param payload - The raw request body as a string
   * @param signature - The X-Webhook-Signature header value
   * @param secret - Your webhook endpoint secret
   * @param timestamp - The X-Webhook-Timestamp header value
   * @param tolerance - Maximum age of the webhook in seconds (default: 300 = 5 minutes)
   *
   * @example
   * ```typescript
   * import { WebhooksModule } from '@tokenisation/sdk';
   * import crypto from 'crypto';
   *
   * app.post('/webhook', (req, res) => {
   *   const isValid = WebhooksModule.verifySignature(
   *     JSON.stringify(req.body),
   *     req.headers['x-webhook-signature'],
   *     process.env.WEBHOOK_SECRET,
   *     parseInt(req.headers['x-webhook-timestamp'])
   *   );
   *
   *   if (!isValid) {
   *     return res.status(401).send('Invalid signature');
   *   }
   *
   *   // Process the webhook...
   *   res.status(200).send('OK');
   * });
   * ```
   */
  static verifySignature(
    payload: string,
    signature: string,
    secret: string,
    timestamp: number,
    tolerance: number = 300
  ): boolean {
    // Check timestamp is within tolerance
    const now = Math.floor(Date.now() / 1000);
    if (Math.abs(now - timestamp) > tolerance) {
      return false;
    }

    // Compute expected signature
    const signedPayloadStr = `${timestamp}.${payload}`;

    // Use dynamic import for crypto in browser/node compatibility
    if (typeof window === 'undefined') {
      // Node.js environment
      const crypto = require('crypto');
      const hmac = crypto.createHmac('sha256', secret);
      hmac.update(signedPayloadStr);
      const expectedSignature = `v1=${hmac.digest('hex')}`;
      return signature === expectedSignature;
    } else {
      // Browser environment - would need SubtleCrypto
      // For now, throw an error as this should be done server-side
      throw new Error('Webhook signature verification should be done server-side');
    }
  }
}
